### PR TITLE
ipv6 support for bnetserver and worldserver

### PR DESCRIPTION
**Pull request by TonyHoyle**
This is ipv6 support.  It works on my system and should be OK on non-ipv6 systems - I've written it such that it avoids touching ipv6 code if it's not configured to do so (to avoid nasties like partially configured routing and such).

To enable ipv6 on the worldserver, just set address6 in auth.realmlist (and of course enable it on the client).

There's also support for ipv6 the bnetserver although the current blizzzard client always connects over ipv4 for this even when a good ipv6 connection is available.. I anticipate they'll add it soon anyway and it's not many extra lines.

To enable it in the bnetserver change the BindIP to '::' or '::1' depending on whether you're using a global or loopback address.

Todo: 
- Test on machines with no ipv6/no v6 routing.  I've tried to code defensively but there's no substitute for a proper test
- Test on Windows... unfortunately I have no easy way to compile on that platform.

### CODE REVIEW

File: sql/updates/auth/2015_09_02_00_auth.sql [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+        AND column_name = 'address6'
+    ) > 0,
+    "SELECT 1",
+    "ALTER TABLE realmlist ADD address6 VARCHAR(255) NOT NULL DEFAULT ''"
+));
+
+PREPARE stmt FROM @s;
+EXECUTE stmt;
+DEALLOCATE PREPARE stmt;
+
```
Comment (DDuarte) [inline sql/updates/auth/2015_09_02_00_auth.sql]: Any reason not to execute the query directly? Also missing change in auth_database.sql file

File: src/server/bnetserver/Realms/RealmList.cpp [hunk 2]
```diff
@@ -119,41 +120,56 @@ void RealmList::UpdateRealms(boost::system::error_code const& error)
                 }
 
                 ip::address externalAddress = (*endPoint).endpoint().address();
+              
+		ip::address externalAddress6;
```
Comment (DDuarte) [inline src/server/bnetserver/Realms/RealmList.cpp]: tabs to spaces

**Comment by DDuarte** [comment]
How useful is ipv6 if the client can't support it?

File: sql/updates/auth/2015_09_02_00_auth.sql (hunk 1 above)
Comment (TonyHoyle) [inline sql/updates/auth/2015_09_02_00_auth.sql]: Just in case the script gets run on a database with the column already there - mysql lacks an 'if column exists' syntax.

Comment (DDuarte) [inline sql/updates/auth/2015_09_02_00_auth.sql]: These sqls are not required to be re-appliable.

**Comment by TonyHoyle** [comment]
The client supports it for connection to worldservers  and has done for some time.  I was somewhat surprised it wouldn't do it for the battlenet server but as that's a trivial part of the patch (just the BindIp support) and they're likely to add it soon anyway it doesn't hurt to be ahead.

**Comment by DDuarte** [comment]
Oh, I was misinformed, thanks

File: sql/updates/auth/2015_09_02_00_auth.sql (hunk 1 above)
Comment (TonyHoyle) [inline sql/updates/auth/2015_09_02_00_auth.sql]: OK I'll just simplify it.

File: src/server/shared/Realm/Realm.cpp [hunk 3]
```diff
@@ -42,8 +42,8 @@ ip::tcp::endpoint Realm::GetAddressForClient(ip::address const& clientAddr) cons
         {
             realmIp = LocalAddress;
         }
-        else
-            realmIp = ExternalAddress;
+        else
```
Comment (tkrokli) [inline src/server/shared/Realm/Realm.cpp]: Is the extra space needed at the end of these 2 lines? Or could they stay as they were?

File: sql/base/auth_database.sql [hunk 4]
```diff
@@ -583,6 +583,7 @@ CREATE TABLE `realmlist` (
   `id` int(10) unsigned NOT NULL AUTO_INCREMENT,
   `name` varchar(32) NOT NULL DEFAULT '',
   `address` varchar(255) NOT NULL DEFAULT '127.0.0.1',
+  `address6` varchar(255) NOT NULL DEFAULT '',
```
Comment (DDuarte) [inline sql/base/auth_database.sql]: Any particular reason this field can't be shared? `address` for both IPV4 and IPV6 addresses

Comment (TonyHoyle) [inline sql/base/auth_database.sql]: The realm server returns both addresses to the client as separate lists, and the client makes the choice (there's a flag in the client under System/Network plus it probably does some local connectivity checks as well) - therefore the realmlist needs to know both.

It could I guess be done as a comma separated list but you have to then ensure there's only one of each and parse the strings rather than just chucking them at boost with the right flag set.

File: src/server/shared/Realm/Realm.cpp (hunk 3 above)
Comment (TonyHoyle) [inline src/server/shared/Realm/Realm.cpp]: That's just vi adding random spaces as it does sometimes.. I'll remove them

**Comment by TonyHoyle** [comment]
Hopefully fixed the email in git this time..

File: sql/base/auth_database.sql (hunk 4 above)
Comment (DDuarte) [inline sql/base/auth_database.sql]: Ah, it's clear now. Two fields are ok then, never been a fan of lists in the DB

**Comment by tkrokli** [comment]
Yes, github is now able to show your user avatar as well as link your username from your commits. :+1:

File: src/server/bnetserver/Realms/WorldListener.cpp [hunk 5]
```diff
@@ -65,7 +68,7 @@ void WorldListener::HandleOpen()
 {
     try
     {
-        _worldSocket->bind(std::string("tcp://*:") + std::to_string(_worldListenPort));
+        _worldSocket->bind(std::string("tcp://"+_bindIp+":") + std::to_string(_worldListenPort));
```
Comment (Shauren) [inline src/server/bnetserver/Realms/WorldListener.cpp]: Can you explain this change? ZMQ sockets are used for ipc, they have nothing to do with clients

File: sql/base/auth_database.sql (hunk 4 above)
Comment (pete318) [inline sql/base/auth_database.sql]: Presumably you want it to bind to both IPv4 and IPv6 interfaces. So best to have the ability to set both addresses.

File: src/server/bnetserver/Realms/WorldListener.cpp (hunk 5 above)
Comment (TonyHoyle) [inline src/server/bnetserver/Realms/WorldListener.cpp]: This simply makes sure that bindIp is honoured.  The old code simply bound to all addresses which is not good on a multihomed server.  Binding always to loopback is also a solution provided the IPC is always local.

It's possible that particular change belongs in its own pull request as it's not directly related to ipv6 - I just noticed it when I was going through the various bits of networking code.

Comment (Shauren) [inline src/server/bnetserver/Realms/WorldListener.cpp]: Indeed it belongs to its own PR.
Also it doesnt have to be always local - bnetserver can be somewhere else.

**Comment by TonyHoyle** [comment]
Tested on a VM with no ipv6 stack, and everything worked normally.

Currently working on the opposite case of an ipv6 only worldserver.

**Comment by Aokromes** [comment]
This branch has conflicts that must be resolved
 Use the command line to resolve conflicts before continuing.

**Comment by tkrokli** [comment]
Not sure if you are aware of this, but you need to rebase your own 6.x branch to solve the conflicts.
Your branch is currently 13 commits behind TrinityCore:6.x . This is mostly due to the new 6.2.2 updates.

**Comment by TonyHoyle** [comment]
It's done on my machine, alas I haven't rebased before and lost a bunch of stuff so I'll have to get back to a working state first.

**Comment by TonyHoyle** [comment]
Tested with an ipv6 only worldserver.. works OK.  Also did some cleaning up of the patch (trailing spaces mostly).

File: sql/updates/auth/2015_09_02_00_auth.sql (hunk 1 above)
Comment (TonyHoyle) [inline sql/updates/auth/2015_09_02_00_auth.sql]: I put the check back - the bnetserver does try to reapply these files if it thinks they've changed, and it was stopping me starting the server at all...

Comment (DDuarte) [inline sql/updates/auth/2015_09_02_00_auth.sql]: Don't. That re-apply on update is a feature. It does not re-apply under normal conditions (once this is added to the repo).

File: src/server/bnetserver/Server/Session.cpp [hunk 6]
```diff
@@ -1141,7 +1144,11 @@ Battlenet::WoWRealm::ListUpdate* Battlenet::Session::BuildListUpdate(Realm const
         version << buildInfo->MajorVersion << '.' << buildInfo->MinorVersion << '.' << buildInfo->BugfixVersion << '.' << buildInfo->Build;
 
         listUpdate->Version = version.str();
-        listUpdate->Address = realm->GetAddressForClient(GetRemoteIpAddress());
+        // Prioritise ipv4, for compatibility
```
Comment (Shauren) [inline src/server/bnetserver/Server/Session.cpp]: This address is always v4 - client reads it as such.

Comment (TonyHoyle) [inline src/server/bnetserver/Server/Session.cpp]: In the case where this happens (ipv6 only worldserver), there isn't an ipv4 address available.  The client seems to handle it.

Comment (Shauren) [inline src/server/bnetserver/Server/Session.cpp]: Because this block is not sent by default.

**Comment by Aokromes** [comment]
This branch has conflicts that must be resolved
 Use the command line to resolve conflicts before continuing.
+ rebase plz.

**Comment by Aokromes** [comment]
@TonyHoyle 
This branch has conflicts that must be resolved
 Use the command line to resolve conflicts before continuing.

If TonyHoyle don't answers, anyone can takeover this work?

**Comment by jackpoz** [comment]
This feature will be implemented sometime in the future by Shauren, taking care of merge conflicts with current sources. Thank you for the contribution.
